JoepegAuctionHouse

0xnivek opened the pull request:
This PR proposes a `JoepegAuctionHouse.sol` which supports running **English and Dutch Auctions on ERC721 tokens**. I'm putting up this PR to get feedback while I work on:

- [x] Tests
- [x] Deploy scripts

To see examples of how auctions are run on other marketplaces, see the **References** section of https://www.notion.so/traderjoexyz/Create-auction-bid-flow-dd5bdb8cacb34874852eb8c36d4a0275

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,960 @@
+
+    uint256 public immutable PERCENTAGE_PRECISION = 10000;
+
+    address public WAVAX;
```
this sounds good to me but had quick question — I thought upgradeable contracts couldn't have constructors. is this safe to do?

LouisMeMyself replied:
Upgradeable contracts can have constructor, it's just that the values that aren't stored within the bytecode (any values that aren't immutable or constants), will not have the expected value because it will load the proxy's storage slot instead of the implementation ones.
That also means that it can't be changed. For example, if you use clone pattern, all cloned contract would have the same values for the immutable and constant because the constructor would not be run again.

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,960 @@
+    );
+
+    modifier isSupportedCurrency(address _currency) {
+        if (!currencyManager.isCurrencyWhitelisted(_currency)) {
```
yeah `WAVAX` is allowlisted inside `currencyManager` already. I think I prefer to keep logic like this in case there is  a scenario where we want to turn it off for some reason and also just to keep the source of truth to be `CurrencyManager`

LouisMeMyself replied:
Currently the specific AVAX/WAVAX functions doesn't check the currencyManager, so there is no way to turn wavax/avax off

LouisMeMyself replied:
My bad I mixed start and settle functions

LouisMeMyself replied:
My bad I mixed start and settle functions

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,960 @@
+    /// @param _duration number of seconds for English Auction to run
+    /// @param _startPrice minimum starting bid price
+    function startEnglishAuction(
+        address _collection,
```
I kept it as `address` for some callsites b/c I was using it as `address` multiple times in the function body but only needed to convert to `IERC721` once so thought its cleaner this way.
slightly tangential to this, do you think I should add a check in the beginning of the function to ensure that the collection conforms to ERC721 interface (i.e. isn't ERC1155 token)? I believe since I use `IERC721.safeTransferFrom` it will revert if it isnt ERC721 but that happens at the end of the function

0xnivek replied:
makes sense, thanks for the thorough explanation! i've changed them to `IERC721` type in the params

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,960 @@
+        uint256 _tokenId,
+        address _currency,
+        uint256 _duration,
+        uint256 _startPrice
```
yes I think so but we probably don't want to allow bids of 0 so I added a check for `_bidAmount != 0` in the `_placeEnglishAuctionBid` function

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,960 @@
+        _clearEnglishAuction(_collection, _tokenId);
+
+        // Settle auction using latest bid
+        if (auction.currency == WAVAX) {
```
hm thats a good question. maybe we could enforce caller is EOA? @LouisMeMyself do you have any thoughts here?

0xnivek replied:
another idea just in case of emergency is adding an `onlyOwner` function that allows owner to end an auction (i.e. return NFT to seller and bid to bidder)

LouisMeMyself replied:
I just realized that the bidder could also be a contract that can't receive AVAX directly.
Maybe the cancel could have an option to send WAVAX instead of AVAX

0xnivek replied:
hm I see — that's probably why `JoepegExchange` also just sends out WAVAX instead of AVAX. i think i'll do the same to avoid complicating the contract code

0xnivek commented:
thanks for all the thorough feedback @LouisMeMyself @simsk87 @Mathieu-Be ! 🙏  I addressed your feedback + left some replies

LouisMeMyself commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -90,7 +93,7 @@ contract JoepegAuctionHouse is
     ///   bid will extend the auction's end time
     /// - Number of seconds to extend an English Auction's end time by
-    uint256 public englishAuctionRefreshTime;
+    uint96 public englishAuctionRefreshTime;
```
Keep it uint256 as it will use a full storage slot either way.

You can also change `englishAuctionMinBidIncrementPct ` to be both `uint128`. If so they will be packed together saving some gas when loading them together in a function

0xnivek replied:
I think its better to keep this as `uint96` b/c it helps me to avoid having to cast from `uint256` -> `uint96` in `_placeEnglishAuctionBid` on line 729 where I increment `auction.endTime` (uint96) by `englishAuctionRefreshTime`

LouisMeMyself commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -250,35 +270,45 @@ contract JoepegAuctionHouse is
-        _placeEnglishAuctionBid(_collection, _tokenId, _currency, _amount);
+    ) external nonReentrant {
+        EnglishAuction memory auction = englishAuctions[_collection][_tokenId];
+        address currency = auction.currency;
+        if (currency == address(0)) {
+            revert JoepegAuctionHouse__NoAuctionExists();
+        }
```
This check i redundant with L712, but I'm fine over checking things, up to you

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -43,28 +45,29 @@ contract JoepegAuctionHouse is
-        uint256 endTime;
-        uint256 dropInterval;
+        uint32 dropInterval;
     }
```
ah I see! I'm going to use 3 `uint256` then — mainly b/c `JoepegExchange` uses `uint256` for prices so I want to be consistent here

0xnivek commented:
Update:
- Changed contract to send out WAVAX instead of AVAX to avoid edge cases where the receiver isn't an EOA and can't receive AVAX
- Added tests
- Added deploy script

Mathieu-Be commented on deploy/JoepegAuctionHouse.js:
```diff
@@ -0,0 +1,79 @@
+    });
+  });
+
+  await verify(proxyContract.implementation, constructorArgs);
```
You should add a check here that you're not on Hardhat because iirc it crashes if you're deploying on your local network

Mathieu-Be reviewed (approved):
LGTM

LouisMeMyself commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,1154 @@
+                uint256 royaltyFeeAmount
+            ) = royaltyFeeManager.calculateRoyaltyFeeAndGetRecipient(
+                    _collection,
+                    _tokenId,
+                    _amount
+                );
+
+            // Check if there is a royalty fee and that it is different to 0
+            if (
+                (royaltyFeeRecipient != address(0)) && (royaltyFeeAmount != 0)
+            ) {
+                IERC20(WAVAX).safeTransfer(
+                    royaltyFeeRecipient,
+                    royaltyFeeAmount
+                );
+                finalSellerAmount -= royaltyFeeAmount;
+
+                emit RoyaltyPayment(
+                    _collection,
+                    _tokenId,
+                    royaltyFeeRecipient,
+                    address(WAVAX),
+                    royaltyFeeAmount
+                );
+            }
+        }
+
+        // Ensure seller gets minimum expected fees
+        if (
+            finalSellerAmount * PERCENTAGE_PRECISION <
+            _minPercentageToAsk * _amount
+        ) {
+            revert JoepegAuctionHouse__FeesHigherThanExpected();
+        }
+
+        // 3. Transfer final amount (post-fees) to seller
+        {
+            IERC20(WAVAX).safeTransfer(_to, finalSellerAmount);
+        }
+    }
```
These 2 functions are now identical if you use WAVAX as the currency in the `_transferFeesAndFunds` function. The WAVAX one should be removed to simplify the contract

LouisMeMyself reviewed (approved):
LGTM

LouisMeMyself commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,1154 @@
+
+    /// @notice Stores English Auction data for NFTs
+    /// @dev (collection address => token id => english auction)
+    mapping(address => mapping(uint256 => EnglishAuction))
+        public englishAuctions;
```
(This is not very important, but it's best code practice)

0xnivek commented on contracts/JoepegAuctionHouse.sol:
```diff
@@ -0,0 +1,1154 @@
+        isValidMinPercentageToAsk(_minPercentageToAsk)
+        nonReentrant
+    {
+        if (_duration == 0 || _duration < _dropInterval) {
```
nope! if `_dropInterval == 0`, that means there is no price drop over time in which case it prob doesnt make sense for it to be a dutch auction

simsk87 replied:
do we check this anywhere then?

0xnivek replied:
o good catch will add

0xnivek replied:
done!

LouisMeMyself reviewed (approved):
LGTM
